wordpress.org/public_html/wp-content/plugins/wporg-profiles-wp-activity-notifier/wporg-profiles-wp-activity-notifier.php

```diff
@@ -150,6 +151,71 @@ public function maybe_notify_new_published_post( $new_status, $old_status, $post
 		$this->notify_new_blog_post( $post );
 	}
 
+	public function insert_post ($postId, $post) {
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			return;
+		}
+		
+		if ( ! $this->is_post_notifiable( $post ) ) {
+			return;
+		}
+
+		if ( $post->post_type === 'wporg_workshop' ) {
+			$this->notify_workshop_presenter( $post );
+		}
+	}
+
+	/**
+	 * Sends activity notification for workshop presenter.
+	 *
+	 * @param WP_Post $post The published post
+	 */
+	public function notify_workshop_presenter( $post ) {
+		if ( defined( 'WP_IMPORTING' ) && WP_IMPORTING ) {
+			return;
+		}
+
+		$presenter_wporg_username = get_post_meta( $post->ID, 'presenter_wporg_username');
+
+		if ( empty( $presenter_wporg_username ) ) {
+			return;
+		}
+		
+		$permalink = get_permalink( $post );
+
+		$title = wp_kses_data( $post->post_title );
+		
+		$content = wp_trim_words(
+			strip_shortcodes( has_excerpt( $post ) ? $post->post_excerpt : $post->post_content ),
+			55
+		);
+
+		foreach ($presenter_wporg_username as $username) {
+			
+			sleep(1);
```

Comment:
It seems we got the rate limit, so sleep a second before the next call out.
(For the case that author and presenter are the same person)

Reply:
I don't think `sleep()` is a good solution in production, though it's useful for diagnosing the problem. It could theoretically lead to unexpected UI delays when publishing a post. Better to find a more robust solution: either find out why the timeout is occurring and solve that, or schedule the function to run in a cron job.

Reply:
I should remove this. It was added because at first I was putting the logic in `transition_post_status`.
Now that the logic is under `wp_insert_post`, the problem of writing data to DB twice in a short time no longer exists.
(It would write the `Wrote a new workshop` activity first during `transition_post_status` first and then after the post is published, then the `Assigned presenter` is sent to DB. Previously, these two would be fired almost at the same time.)

I'm still curious what's the practice for us to handle such case (write DB in a row in a short time) since the logic might be moved back to use `transition_post_status`.
cron job sounds like a good idea.

> either find out why the timeout is occurring and solve that
What does timeout here refer to, timeout from sleep? or timeout of saving to DB? Thanks!

Reply:
>> either find out why the timeout is occurring and solve that
> What does timeout here refer to

I believe Alex meant rate-limit here, not timeout.

There shouldn't be any rate-limits in production that would catch this - if there is, we should probably get those lifted for self-calls.

Reply:
Yep sorry Dion's right, I meant the rate limit.